### PR TITLE
Ins 62 make wallet smart contract

### PR DESCRIPTION
**Pull request by ivanshibitov**
**- What I did**
1) move Composite* from object to factory
2) add wallet smart contract
3) add tests

**- How to verify it**
Run tests

### CODE REVIEW

File: genesis/example/wallet.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/insolar/insolar/genesis/model/contract"
+	"github.com/insolar/insolar/genesis/model/factory"
+	"github.com/insolar/insolar/genesis/model/object"
+	"github.com/insolar/insolar/genesis/model/resolver"
+)
+
+type Wallet interface {
+	factory.Composite
+	contract.SmartContract
+
```
Comment (megge-dream) [inline genesis/example/wallet.go]: Space here is not very pleasant to the eye.

File: genesis/example/wallet.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	GetBalance() int
+}
+
+type wallet struct {
+	contract.BaseSmartContract
+
```
Comment (megge-dream) [inline genesis/example/wallet.go]: Space here is not very pleasant to the eye.

File: genesis/example/wallet.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+func (w *wallet) GetClassID() string {
+	return class.WalletID
+}
+
+func (w *wallet) GetInterfaceKey() string {
+	return w.GetClassID()
+}
+
+func newWallet(parent object.Parent) (Wallet, error) {
```
Comment (megge-dream) [inline genesis/example/wallet.go]: newWallet func should be placed right after wallet struct

File: genesis/example/wallet.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+func (wf *walletFactory) GetInterfaceKey() string {
+	return class.WalletID
+}
+
+func (wf *walletFactory) GetClassID() string {
+	return class.WalletID
+}
+
+func (*walletFactory) Create(parent object.Parent) (factory.Composite, error) {
```
Comment (megge-dream) [inline genesis/example/wallet.go]: Naming *walletFactory even than it does not used in method can be more readable. What do you think?

File: genesis/example/wallet_domain.go [hunk 5]
```diff
@@ -34,59 +34,63 @@ const WalletDomainName = "WalletDomain"
...
-		return nil, err
+		return err
 	}
 
-	result, ok := wallet.(resolver.Proxy)
+	// Check if it Factory
+	wf, ok := child.(factory.CompositeFactory)
 	if !ok {
-		return nil, fmt.Errorf("object with record `%s` is not `Proxy` instance", record)
+		return fmt.Errorf("child by reference `%s` is not Factory instance", wd.walletFactoryReference)
```
Comment (megge-dream) [inline genesis/example/wallet_domain.go]: is not **_CompositeFactory_** instance

File: genesis/example/doc.go [hunk 6]
```diff
@@ -37,5 +37,25 @@ Entities:
 		factory := NewMemberFactory(memberDomain)
 		m, error := factory.Create(parent)
 
+////
+
+    WalletDomain - domain that allows to add new wallets to system
+
+	Usage:
```
Comment (megge-dream) [inline genesis/example/doc.go]: Fix indents

File: genesis/example/wallet.go (hunk 4 above)
Comment (ivanshibitov) [reply genesis/example/wallet.go]: I don't know. But yes,  we name them everywhere

File: genesis/example/doc.go [hunk 7]
```diff
@@ -37,5 +37,25 @@ Entities:
...
+
+		factory := NewWalletDomainFactory( factoryParent )
```
Comment (megge-dream) [inline genesis/example/doc.go]: Loose spaces around param

File: genesis/example/doc.go [hunk 8]
```diff
@@ -37,5 +37,25 @@ Entities:
...
+		wDomain, err := factory.Create( domainParent )
+
+		// since Create inject composite to member proxy
+		err := wDomain.CreateWallet( memberProxy )
+
+	Wallet - smart contract that represent wallet
+
+	Usage:
+		factory := NewWalletFactory(walletDomain)
+		w, err := factory.Create(parent)
```
Comment (megge-dream) [inline genesis/example/doc.go]: We didn't add any examples of working with wallet as with composite. Should be provide it?

File: genesis/example/member_test.go [hunk 9]
```diff
@@ -48,7 +49,26 @@ func (c *BaseComposite) SetReference(reference object.Reference) {}
 
 type BaseCompositeFactory struct{}
```
Comment (ilyapt) [inline genesis/example/member_test.go]: Maybe better name this mock like mockBaseCompositeFactory?

File: genesis/example/wallet.go (hunk 4 above)
Comment (megge-dream) [reply genesis/example/wallet.go]: I added task about deleting such things everywhere to backlog. Here leave it as it is.

File: genesis/example/doc.go (hunk 8 above)
Comment (ivanshibitov) [reply genesis/example/doc.go]: add example for getting user's balance

File: genesis/example/wallet_domain.go [hunk 10]
```diff
@@ -34,59 +34,63 @@ const WalletDomainName = "WalletDomain"
...
-}
-
-func (wd *walletDomain) GetClassID() string {
-	return class.WalletDomainID
-}
+	}
 
-func (wd *walletDomain) CreateWallet(fc factory.Factory) (string, error) {
-	wallet, err := fc.Create(wd)
+	// Add memberFactory as a child
```
Comment (megge-dream) [inline genesis/example/wallet_domain.go]: _Add **walletFactory** as a child_

**Pull request merged**